game/dino_jump_main.py

```diff
@@ -0,0 +1,394 @@
+        # Using timing to determine if we are mid jump and y position should be adjusted
+        if pygame.time.get_ticks() - self.jump_start_time <= self.JUMP_TIME:
+            jump_time = pygame.time.get_ticks() - self.jump_start_time
+            self.y = self.OG_DINO_Y - self.JUMP_COEF * (
+                jump_time**2 - self.JUMP_TIME * jump_time
+            )
+        else:
+            if self.is_jumping:
+                self.jump_start_time = pygame.time.get_ticks()
+            else:
+                self.y = self.OG_DINO_Y
+
+    def update_frame(self):
+
+        # Determine when the dino foot change happens
+        if self.curr_walk <= 0:
+            self.is_right_foot_down = not self.is_right_foot_down
+            self.curr_walk = self.UPDATE_WALK_FRAMES
+        self.curr_walk -= 1
+
+        # Update dino pixel array
+        if (
+            self.is_jumping
+            or pygame.time.get_ticks() - self.jump_start_time < self.JUMP_TIME
+        ):
+            self.ARRAY = self.JUMP_ARRAY
+        elif self.is_ducking and self.is_right_foot_down:
+            self.ARRAY = self.RDUCK_ARRAY
+        elif self.is_ducking:
+            self.ARRAY = self.LDUCK_ARRAY
+        elif self.is_right_foot_down:
+            self.ARRAY = self.RRUN_ARRAY
+        else:
+            self.ARRAY = self.LRUN_ARRAY
+
+        # If the game is over, show the dead dino frame
+        if game.game_over:
+            self.ARRAY = self.DONE_ARRAY  # JOSH a new sprite array for the "done" look
+
+
+#############################################################################
+# VOID SETUP
+#############################################################################
+
+# Initialises the Dino Game
+game = Game()
+
+# Start Pygame
+pygame.init()
+game.draw_all()
+
+
+#############################################################################
+# VOID LOOP
+#############################################################################
+
+# Wait until first space bar press until the game starts
+while not game.start_game:
+    game.check_start_game()
+
+# Simulates the whole loop running
+while not game.game_over:
+    game.check_user_input()
+    game.dino.update_jumping_location()
+    game.all_obstacles.generate_obstacle()
+    game.dino.update_frame()
+    game.all_obstacles.update_location()
+    game.draw_all()
+
+print("game over!")  # JOSH everything after here is the end game state!
+game.dino.update_frame()
+game.draw_all()
+while not game.close_all:
+    game.check_close_panel()
+
+
+# CURRENT IMPLEMENTATION: DRAW_AND_CHECK_COLLISION_ALL
+# Ideally, draw and check_collision would be two separate functions
+# However, currently running the two functions separately results in the game
+# ending much earlier than expected
+# Current Rationale: Since the function determines the set_location twice for one obs,
+# it could affect future calculations
+# Need to see how the data within the obstacles are changed after running the function.

```

Comment:
Personally I would avoid checking these comments in, but if you want to keep track of what improvements/further implementations are needed I would put them as TODO comments.

Reply:
I'll keep some of the comments locally, removed from the PR.